### PR TITLE
autoconf 2.71 (+ automake update)

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

`autoconf` now needs to depend on brewed `m4`, or else `automake`'s configure script will complain about a broken `autoconf`.

Alternatively, we could just depend on brewed `m4` only for the `automake` build in order to mislead `automake`'s `configure` script, but that could lead to other broken things in case `automake`'s test for a working `autoconf` was not a false positive. (I haven't actually tested this, but I _think_ it should work.)

Follow up to #66511. See also #70617.

-----

### TODO

- [x] Re-run CI when https://github.com/Homebrew/brew/pull/10921 is merged.
- [x] Merge when `brew 3.0.10` or `brew 3.1.0` is tagged.

### CODE REVIEW

**Comment by carlocab** [comment]
It looks like superenv forces `autoconf` to try to use the macOS built-in `m4` [here](https://github.com/Homebrew/brew/blob/00d8e1ab3dab7644f93fccdfbabb4e24310d24aa/Library/Homebrew/extend/ENV/super.rb#L77):
```ruby
    self["M4"] = DevelopmentTools.locate("m4") if deps.any? { |d| d.name == "autoconf" }
```
We could either:

1. keep that, and just trick `automake`'s `configure` script here by using brewed `m4` just for the `automake` build (and `autoconf` probably no longer needs to depend on `m4`); or,
2. change the above, in case `automake` was right to complain about the built-in macOS `m4`.

`autoconf` actually builds and tests fine without brewed `m4`; it only fails when you try to run `#{pkgshare}/autotest/autotest.m4`, which looks a lot like what `automake` is doing in its `configure` script.

Thoughts, @Bo98, @fxcoudert?

File: Formula/automake.rb [hunk 1]
```diff
@@ -15,20 +16,10 @@ class Automake < Formula
 
   depends_on "autoconf"
 
-  # Download more up-to-date config scripts.
-  resource "config" do
-    url "https://git.savannah.gnu.org/cgit/config.git/snapshot/config-0b5188819ba6091770064adf26360b204113317e.tar.gz"
-    sha256 "3dfb73df7d073129350b6896d62cabb6a70f479d3951f00144b408ba087bdbe8"
-    version "2020-08-17"
-  end
```
Comment (Bo98) [inline Formula/automake.rb:23]: Can you explain why this is no longer needed?

Comment (carlocab) [reply Formula/automake.rb:23]: I don't know, to be honest. I just inherited that from https://github.com/Homebrew/homebrew-core/pull/66511/commits/82f7491cdf202c2e8c7bab26071c7d6f64f9b8fb

Comment (Bo98) [reply Formula/automake.rb:23]: You can check what `config.guess`/`config.sub` version is installed. If it's older than 2020-08-17 then this is still needed.

Comment (carlocab) [reply Formula/automake.rb:23]: From `config.guess`:

```sh
#! /bin/sh
# Attempt to guess a canonical system name.
#   Copyright 1992-2020 Free Software Foundation, Inc.

timestamp='2020-11-07'
```

Same in `config.sub`. Is that adequate?

Comment (Bo98) [reply Formula/automake.rb:23]: Ah yes, it was included in automake 1.16.3 but the resource wasn't properly removed at the time. The autoconf relation confused me there - it's actually completely unrelated.

**Comment by Bo98** [comment]
> We could either:
> 
> 1. keep that, and just trick automake's configure script here by using brewed m4 just for the automake build (and autoconf probably no longer needs to depend on m4); or,
> 2. change the above line in superenv, in case automake was right to complain about the built-in macOS m4.

That's an interesting debate actually. I know licenses haven't traditionally been too much of a concern, but using brewed `m4` everywhere would mean anyone with `HOMEBREW_FORBIDDEN_LICENSES=GPL-3.0` can no longer build formulae using autotools.

**Comment by carlocab** [comment]
Well, there's also the question of whether `automake`'s configure script is wrong when it says `autoconf` + macOS `m4` is broken. If it isn't, then I guess we don't really have much of a choice on brewed `m4` everywhere.

**Comment by carlocab** [comment]
> That's an interesting debate actually. I know licenses haven't traditionally been too much of a concern, but using brewed `m4` everywhere would mean anyone with `HOMEBREW_FORBIDDEN_LICENSES=GPL-3.0` can no longer build formulae using autotools.

Actually, `autoconf` is `GPL-3.0-or-later` now, so it seems we don't really have a choice on the licensing issue:

```
❯ autoconf --version
autoconf (GNU Autoconf) 2.71
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+/Autoconf: GNU GPL version 3 or later
<https://gnu.org/licenses/gpl.html>, <https://gnu.org/licenses/exceptions.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David J. MacKenzie and Akim Demaille.
```

**Comment by Bo98** [comment]
In that case, ignore what I said about that.

**Comment by Bo98** [comment]
So I looked into this. This stems from a bug in `m4` where it loses track of the current file and line under some situations. I'm not entirely sure when it was fixed. 1.4.8 revamped how line tracking worked so it could well be that version fixed it (I still don't know why Apple's version doesn't have these changes since the GPLv3 move happened in 1.4.10).

You can see the bug here:

```
$ /usr/bin/m4 --nesting-limit=1024 --include=/usr/local/Cellar/autoconf/2.71/share/autoconf --debug=aflqi --fatal-warning --trace=_m4_warn --reload-state=/usr/local/Cellar/autoconf/2.71/share/autoconf/autoconf/autoconf.m4f /usr/local/Cellar/autoconf/2.71/share/autoconf/autoconf/trailer.m4 conftest.ac > /dev/null
m4debug: input read from /usr/local/Cellar/autoconf/2.71/share/autoconf/autoconf/trailer.m4
m4debug:/usr/local/Cellar/autoconf/2.71/share/autoconf/autoconf/trailer.m4:4: input exhausted
m4debug: input read from conftest.ac
m4debug:conftest.ac:1: input read from /usr/local/Cellar/autoconf/2.71/share/autoconf/m4sugar/foreach.m4
m4debug:/usr/local/Cellar/autoconf/2.71/share/autoconf/m4sugar/foreach.m4:362: input reverted to conftest.ac, line 1
m4debug:conftest.ac:1: input exhausted
m4trace: -1- _m4_warn([syntax], [AC_OUTPUT was never used], [])
```

```
$ /usr/local/opt/m4/bin/m4 --nesting-limit=1024 --include=/usr/local/Cellar/autoconf/2.71/share/autoconf --debug=aflqi --fatal-warning --trace=_m4_warn --reload-state=/usr/local/Cellar/autoconf/2.71/share/autoconf/autoconf/autoconf.m4f /usr/local/Cellar/autoconf/2.71/share/autoconf/autoconf/trailer.m4 conftest.ac > /dev/null
m4debug: input read from /usr/local/Cellar/autoconf/2.71/share/autoconf/autoconf/trailer.m4
m4debug:/usr/local/Cellar/autoconf/2.71/share/autoconf/autoconf/trailer.m4:5: input exhausted
m4debug: input read from conftest.ac
m4debug:conftest.ac:1: input read from /usr/local/Cellar/autoconf/2.71/share/autoconf/m4sugar/foreach.m4
m4debug:/usr/local/Cellar/autoconf/2.71/share/autoconf/m4sugar/foreach.m4:363: input reverted to conftest.ac, line 1
m4debug:conftest.ac:2: input exhausted
m4trace:/usr/local/Cellar/autoconf/2.71/share/autoconf/autoconf/trailer.m4:4: -1- _m4_warn([syntax], [AC_OUTPUT was never used], [])
```

(The last line in particular)

It's mostly as a result of `m4_wrap`, where autoconf's trailer.m4 does a `m4_wrap([_AC_FINALIZE])` to dynamically insert a `_AC_FINALIZE` at the end of all inputs. This seems to trigger the bug in older m4 versions.

This is why other warnings are not affected. For example, the `AC_USG` deprecation warning works fine because that warning is thrown on that same line (the call to `AC_USG`) and has nothing to do with `m4_wrap`/`_AC_FINALIZE`. The `AC_OUTPUT` warning is however thrown _inside_ `_AC_FINALIZE`, so that's why the bug is triggered for that warning specifically.

Can it be worked around? Sure. A two-line patch to `autom4te` to handle such bugged traces is possible (I tried it).

Is it worth it? Probably not. I'm seeing minimal benefits of sticking with Apple's `m4` at this point.

**Comment by carlocab** [comment]
Nice, thanks for looking at it.

I just tested whether `autoconf` built against macOS `m4` can still be used to build `automake` as long as brewed `m4` is what's used in the `automake` build, and it seems that that's possible too. However, I'm not sure it's worth it to keep going out of our way to fix this if similar issues arise going forward.

**Comment by Bo98** [comment]
> I just tested whether `autoconf` built against macOS `m4` can still be used to build `automake` as long as brewed `m4` is what's used in the `automake` build, and it seems that that's possible too.

How `autoconf` is built really doesn't matter in the context of the superenv. M4 isn't invoked in the `autoconf` build (beyond some tests whether it is functional or not) - it's used to bake in default values to `autom4te` etc in the absence of a `M4` env, e.g.

```perl
my $m4 = $ENV{"M4"} || '/usr/bin/m4';
```

So we shouldn't do anything hybrid here.

I also suggest checking this isn't using the Cellar path or anything like that.

**Comment by carlocab** [comment]
> I also suggest checking this isn't using the Cellar path or anything like that. I don't think it will be though.

Nope. In `autom4te`:
```perl
# $M4.
my $m4 = $ENV{"M4"} || '/usr/local/opt/m4/bin/m4';
```
and in `autoupdate`:
```perl
# $M4.
my $m4 = $ENV{"M4"} || '/usr/local/opt/m4/bin/m4';
```

The only Cellar references I can find are to `autoconf`'s Cellar path.

**Comment by fxcoudert** [comment]
I'd be for keeping it simple, and depending on `m4`. Apple's version is really old, and they've consistently been telling us (python, ruby, etc) that they don't intend to update those much, and we should rely on our own versions.

File: Formula/automake.rb [hunk 2]
```diff
@@ -15,20 +16,10 @@ class Automake < Formula
...
-
   def install
     on_macos do
       ENV["PERL"] = "/usr/bin/perl"
-    end
-
-    resource("config").stage do
-      cp Dir["config.*"], buildpath/"lib"
+      ENV.delete "M4"
```
Comment (fxcoudert) [inline Formula/automake.rb:22]: Why are we removing `M4` here?

Comment (carlocab) [reply Formula/automake.rb:22]: The build fails if we don't because of superenv.

File: Formula/automake.rb [hunk 3]
```diff
@@ -15,20 +16,12 @@ class Automake < Formula
...
-
   def install
     on_macos do
       ENV["PERL"] = "/usr/bin/perl"
-    end
-
-    resource("config").stage do
-      cp Dir["config.*"], buildpath/"lib"
+      # TODO: Remove when https://github.com/Homebrew/brew/pull/10921
+      # lands in a tagged release of brew
```
Comment (Bo98) [inline Formula/automake.rb:23]: I think this whole PR has to wait for a tagged release, as any usage of autoconf 2.71 in other formulae relies on that being removed.

Comment (carlocab) [reply Formula/automake.rb:23]: I actually think it'll work fine in those cases as long as they don't run `automake`'s test for `autoconf`, but, yes, better to play it safe.

Comment (Bo98) [reply Formula/automake.rb:23]: It's a bit more than that. It affects any autoconf file that has a `AC_INIT` but no `AC_OUTPUT`.

Comment (carlocab) [reply Formula/automake.rb:23]: Aha, ok. I've managed to build some things just fine with `autoconf` + `automake` as in this PR while superenv still set `M4`, but I guess I was just lucky.

**Comment by carlocab** [comment]
Updated `libtool` to also depend on brewed `m4` (so we don't have mixed versions in the dependency tree). More context at #73932.

**Comment by carlocab** [comment]
`libtool` commit moved to #73981.

**Comment by carlocab** [comment]
brew 3.0.10 has been tagged. Merging.

**Pull request closed without merging**